File: Maximum Rectangle in Binary Matrix.c
#include <stdio.h>
#define MAX 100

int maxHist(int row[], int n) {
    int stack[MAX], top = -1, max = 0, i = 0;

    while (i < n) {
        if (top == -1 || row[stack[top]] <= row[i])
            stack[++top] = i++;
        else {
            int h = row[stack[top--]];
            int w = (top == -1) ? i : i - stack[top] - 1;
            if (h * w > max) max = h * w;
        }
    }

    while (top != -1) {
        int h = row[stack[top--]];
        int w = (top == -1) ? i : i - stack[top] - 1;
        if (h * w > max) max = h * w;
    }

    return max;
}

int maxRectangle(int M[][MAX], int R, int C) {
    int result = maxHist(M[0], C);

    for (int i = 1; i < R; i++) {
        for (int j = 0; j < C; j++)
            if (M[i][j])
                M[i][j] += M[i - 1][j];

        int area = maxHist(M[i], C);
        if (area > result) result = area;
    }
    return result;
}

File: Maximum Rectangle in Binary Matrix.py
def largestHistogram(heights):
    stack = []
    max_area = 0
    heights.append(0)

    for i in range(len(heights)):
        while stack and heights[stack[-1]] > heights[i]:
            h = heights[stack.pop()]
            w = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, h * w)
        stack.append(i)

    heights.pop()
    return max_area


def maxRectangle(matrix):
    if not matrix:
        return 0

    cols = len(matrix[0])
    heights = [0] * cols
    max_area = 0

    for row in matrix:
        for i in range(cols):
            heights[i] = heights[i] + 1 if row[i] == 1 else 0
        max_area = max(max_area, largestHistogram(heights))

    return max_area
